actors/builtin/power/power_actor.go

```diff
@@ -403,6 +442,85 @@ func (a Actor) computeInitialPledge(rt Runtime, desc *SectorStorageWeightDesc) a
 	return initialPledge
 }
 
+func (a Actor) processBatchProofVerifies(rt Runtime) error {
+	var st State
+
+	var miners []address.Address
+	verifies := make(map[address.Address][]abi.SealVerifyInfo)
+
+	rt.State().Transaction(&st, func() interface{} {
+		store := adt.AsStore(rt)
+		mmap, err := adt.AsMultimap(store, st.ProofValidationBatch)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to load proofs validation batch: %s", err)
+		}
+
+		err = mmap.ForAll(func(k string, arr *adt.Array) error {
+			a, err := address.NewFromBytes([]byte(k))
+			if err != nil {
+				return xerrors.Errorf("failed to parse address key: %w", err)
+			}
+
+			miners = append(miners, a)
+
+			var infos []abi.SealVerifyInfo
+			var svi abi.SealVerifyInfo
+			err = arr.ForEach(&svi, func(i int64) error {
+				infos = append(infos, svi)
+				return nil
+			})
+			if err != nil {
+				return xerrors.Errorf("failed to iterate over proof verify array for miner %s: %w", a, err)
+			}
+			verifies[a] = infos
+			return nil
+		})
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to iterate proof batch: %s", err)
+		}
+
+		emptyCid, err := adt.MakeEmptyMultimap(store).Root()
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to get empty multimap cid")
+		}
+		st.ProofValidationBatch = emptyCid
+
+		return nil
+	})
+
+	res, err := rt.Syscalls().BatchVerifySeals(verifies)
+	if err != nil {
+		rt.Abortf(exitcode.ErrIllegalState, "failed to batch verify: %s", err)
+	}
+
+	for _, m := range miners {
+		vres, ok := res[m]
+		if !ok {
+			rt.Abortf(exitcode.ErrNotFound, "batch verify seals syscall implemented incorrectly")
+		}
+
+		verifs := verifies[m]
+
+		var successful []abi.SectorNumber
+		for i, r := range vres {
+			if r {
+				successful = append(successful, verifs[i].SectorID.Number)
+			}
+		}
+
+		ret, code := rt.Send(
+			m,
+			builtin.MethodsMiner.ConfirmSectorProofsValid,
+			&builtin.ConfirmSectorProofsParams{successful},
+			abi.NewTokenAmount(0),
+		)
+		builtin.RequireSuccess(rt, code, "failed to confirm sector proofs valid") // should never happen...
+		_ = ret
```

Comment:
There's a fairly large amount of abort cases in that call, I'm not 100% convinced that this can't break things in some way

Reply:
Yeah, i'm slightly concerned about it, I'll see if i can clean that up.

Reply:
Even if we're pretty sure `ConfirmSectorProofsValid` won't abort, I think we should be defensive here and let one miner fail without aborting this whole thing, which will bork all commitments forever since state will roll back.